README.md
[![Sensu Bonsai Asset](https://img.shields.io/badge/Bonsai-Download%20Me-brightgreen.svg?colorB=89C967&logo=sensu)](https://bonsai.sensu.io/assets/sensu/sensu-sumologic-handler)
![goreleaser](https://github.com/sensu/sensu-sumologic-handler/workflows/goreleaser/badge.svg)
[![Go Test](https://github.com/sensu/sensu-sumologic-handler/workflows/Go%20Test/badge.svg)](https://github.com/sensu/sensu-sumologic-handler/actions?query=workflow%3A%22Go+Test%22)
[![goreleaser](https://github.com/sensu/sensu-sumologic-handler/workflows/goreleaser/badge.svg)](https://github.com/sensu/sensu-sumologic-handler/actions?query=workflow%3Agoreleaser)


# Sensu Sumo Logic Handler

## Table of Contents
- [Overview](#overview)
- [Usage](#usage)
- [Environment variables](#environment-variables)
- [Annotations](#annotations)
- [Configuration](#configuration)
  - [Asset registration](#asset-registration)
  - [Handler definition](#handler-definition)
- [Installation from source](#installation-from-source)
- [Additional notes](#additional-notes)
- [Contributing](#contributing)

## Overview

The Sensu Sumo Logic Handler is a [Sensu Handler](https://docs.sensu.io/sensu-go/latest/reference/handlers/) that sends Sensu observability data (events and metrics) to a Sumo Logic [HTTP Logs and Metrics Source](https://help.sumologic.com/03Send-Data/Sources/02Sources-for-Hosted-Collectors/HTTP-Source).
This handler will send the Sensu event as either a log entry, or as a set of metrics or both (see `--send-log`, `--send-metrics` to select the mode of operation).

## Usage

```
Send Sensu observability data (events and metrics) to a hosted Sumo Logic HTTP Logs and Metrics Source.

Usage:
  sensu-sumologic-handler [flags]
  sensu-sumologic-handler [command]

Available Commands:
  help        Help about any command
  version     Print the version number of this plugin

Flags:
  -u, --url string                 Sumo Logic HTTP Logs and Metrics Source URL (Required)
  -l, --send-log                   Send event as log
  -m, --send-metrics               Send event metrics, if there are metrics attached to sensu event
      --log-fields string          Custom Sumo Logic log fields (comma separated key=value pairs)
      --metric-dimensions string   Custom Sumo Logic metric dimensions (comma separated key=value pairs)
      --source-category string     Custom Sumo Logic source category (supports handler templates) (default "sensu-event")
      --source-host string         Custom Sumo Logic source host (supports handler templates) (default "{{ .Entity.Name }}")
      --source-name string         Custom Sumo Logic source name (supports handler templates) (default "{{ .Check.Name }}")
  -n, --dry-run                    Dry-run, do not send data to Sumo Logic collector, report to stdout instead
  -v, --verbose                    Verbose output to stdout
  -h, --help                       help for sensu-sumologic-handler

Use "sensu-sumologic-handler [command] --help" for more information about a command.
```

## Environment variables

|Argument             |Environment Variable         |
|---------------------|-----------------------------|
|--url                |SUMOLOGIC_URL                |
|--send-log           |SUMOLOGIC_SEND_LOG           |
|--send-metrics       |SUMOLOGIC_SEND_METRICS       |
|--source-name        |SUMOLOGIC_SOURCE_NAME        |
|--source-host        |SUMOLOGIC_SOURCE_HOST        |
|--source-category    |SUMOLOGIC_SOURCE_CATEGORY    |
|--metric-dimensions  |SUMOLOGIC_METRIC_DIMENSIONS  |
|--log-fields         |SUMOLOGIC_LOG_FIELDS         |

**Security Note:** Care should be taken to not expose the `--url` for this handler by specifying it on the command line or by directly setting the environment variable in the handler definition.
It is suggested to make use of [secrets management](https://docs.sensu.io/sensu-go/latest/operations/manage-secrets/secrets/) to surface it as an environment variable.
The example [handler definition](#handler-definition) below references the Sumo Logic URL as a secret.
Here is corresponding secret definition that make use of the built-in [env secrets provider](https://docs.sensu.io/sensu-go/latest/operations/manage-secrets/secrets-providers/#env-secrets-provider-example).

```yml
---
type: Secret
api_version: secrets/v1
metadata:
  name: sumologic_url
spec:
  provider: env
  id: SUMOLOGIC_URL
```

## Annotations

All of the command line arguments referenced in the help usage message can be overridden by check or entity annotations.
The annotation consists of the key formed by appending the "long" argument specification to the string `sensu.io/plugins/sumologic/config` (e.g. `sensu.io/plugins/sumologic/config/source-name`).

For example, having the following in an `agent.yml` file will create an entity annotation such that Sensu metrics sent to SumoLogic from this entity will include the additional metric-dimensions string `environment=production, entity=test` instead of the dimensions string defined with the handler command flag.

```yml
namespace: "default"
subscriptions:
  - linux
backend-url:
  - "ws://127.0.0.1:8081"
annotations:
  sensu.io/plugins/sumologic/config/metric-dimensions: "environment=production, entity=test"
```

## Configuration

### Asset registration

[Sensu Assets](https://docs.sensu.io/sensu-go/latest/reference/assets/) are the best way to make use of this plugin.
If you're not using an asset, please consider doing so!
If you're using sensuctl 5.13 with Sensu Backend 5.13 or later, you can use the following command to add the asset:

```
sensuctl asset add sensu/sensu-sumologic-handler
```

If you're using an earlier version of sensuctl, you can find the asset on the [Bonsai Asset Index](https://bonsai.sensu.io/assets/sensu/sensu-sumologic-handler).

### Handler definition

```yml
---
type: Handler
api_version: core/v2
metadata:
  name: sumologic
spec:
  command: >-
    sensu-sumologic-handler --send-log --send-metrics
    --source-host "{{ .Entity.Name }}"
    --source-name "{{ .Check.Name }}"
  type: pipe
  runtime_assets:
  - sensu/sensu-sumologic-handler
  secrets:
  - name: SUMOLOGIC_URL
    secret: sumologic_url
---
type: Secret
api_version: secrets/v1
metadata:
  name: sumologic_url
spec:
  provider: env
  id: SUMOLOGIC_HTTP_COLLECTOR_URL
```
**Note:**  This handler enabled both sending the Sensu event as a log and also sending metrics.
  
#### Proxy Support

This handler supports the use of the environment variables `HTTP_PROXY`, `HTTPS_PROXY`, and `NO_PROXY` (or the lowercase versions thereof).
`HTTPS_PROXY` takes precedence over `HTTP_PROXY` for https requests.
The environment values may be either a complete URL or a `"host[:port]"` value (with no `http://` or `https://` prefix), in which case the "http" scheme is assumed.

## Installation from source

The preferred way of installing and deploying this plugin is to use it as an Asset.
If you would like to compile and install the plugin from source or contribute to it, download the latest version or create an executable script from this source.

From the local path of the sensu-sumologic-handler repository:

```
go build
```

## Additional notes

## Contributing

For more information about contributing to this plugin, see [Contributing](https://github.com/sensu/sensu-go/blob/master/CONTRIBUTING.md).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic
Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased

## [0.3.0] - 2021-11-12

### Breaking Change
- Refactor sensu-event json log output to better match Sumologic http collector's automatic timestamp detection.

## [0.2.0] - 2021-07-26

### Changed
- Refactor log/metric mode selection using new set of cmdline options.
- Remove metric format option, no need to support anything but prometheus
- Modify sensu events passed as logs into http source  to use 13 digit millisecond accuracy timstamp meeting requirements for automatic timestamp field detection.

## [0.1.1] - 2021-07-14

### Fixed
- Fix crash for golang Event object with nil Metrics

## [0.1.0] - 2021-06-25

### Added
- Initial release


main_test.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/http/httptest"
	"net/url"
	"strings"
	"testing"

	corev2 "github.com/sensu/sensu-go/api/core/v2"
	"github.com/stretchr/testify/assert"
)

func contains(s []string, e string) bool {
	for _, a := range s {
		if a == e {
			return true
		}
	}
	return false
}
func clearPlugin() {
	plugin.EnableSendLog = false
	plugin.Url = ""
	plugin.DryRun = false
}

func TestLogMsgTimestampLocation(t *testing.T) {
	event := corev2.FixtureEvent("entity1", "check1")
	expectedTimestamp := msTimestamp(event.Timestamp)
	logMsg, err := createLogMsg(event)
	assert.NoError(t, err)
	data := logMsg.Data[0].(map[string]int64)
	assert.Equal(t, data["timestamp"], expectedTimestamp)
}

func TestCheckArgs(t *testing.T) {
	err := checkArgs(nil)
	assert.Error(t, err)
	plugin.EnableSendLog = true
	err = checkArgs(nil)
	assert.Error(t, err)
	plugin.Url = "test"
	err = checkArgs(nil)
	assert.NoError(t, err)
	clearPlugin()
}

func TestConvertMetric(t *testing.T) {
	nsStamp := int64(1624376039373111122)
	usStamp := int64(1624376039373111)
	msStamp := int64(1624376039373)
	sStamp := int64(1624376039)
	msSecond := int64(1624376039000)
	expectedData := `answer{foo="bar", hey="there"} 42 `
	a := [4]int64{msStamp, usStamp, nsStamp, sStamp}
	event := corev2.FixtureEvent("entity1", "check1")
	event.Check = nil
	event.Metrics = corev2.FixtureMetrics()
	for _, p := range event.Metrics.Points {
		p.Tags = append(p.Tags, &corev2.MetricTag{Name: "hey", Value: "there"})
	}
	for _, stamp := range a {
		for _, p := range event.Metrics.Points {
			p.Timestamp = stamp
		}
		dataString, err := convertMetrics(event)
		assert.NoError(t, err)
		msTime := expectedData + fmt.Sprintf("%v\n", msStamp)
		if stamp < msStamp {
			msTime = expectedData + fmt.Sprintf("%v\n", msSecond)
		}
		usTime := expectedData + fmt.Sprintf("%v\n", usStamp)
		nsTime := expectedData + fmt.Sprintf("%v\n", nsStamp)
		sTime := `answer{foo="bar"} 42 ` + fmt.Sprintf("%v\n", sStamp)
		assert.Equal(t, msTime, dataString)
		assert.NotEqual(t, usTime, dataString)
		assert.NotEqual(t, nsTime, dataString)
		assert.NotEqual(t, sTime, dataString)
	}
}
func TestConvertMetricWithNilTags(t *testing.T) {
	nsStamp := int64(1624376039373111122)
	usStamp := int64(1624376039373111)
	msStamp := int64(1624376039373)
	sStamp := int64(1624376039)
	msSecond := int64(1624376039000)
	a := [4]int64{msStamp, usStamp, nsStamp, sStamp}
	event := corev2.FixtureEvent("entity1", "check1")
	event.Check = nil
	event.Metrics = corev2.FixtureMetrics()
	for _, stamp := range a {
		for _, p := range event.Metrics.Points {
			p.Timestamp = stamp
			p.Tags = nil
		}
		dataString, err := convertMetrics(event)
		assert.NoError(t, err)
		msTime := `answer{} 42 ` + fmt.Sprintf("%v\n", msStamp)
		if stamp < msStamp {
			msTime = `answer{} 42 ` + fmt.Sprintf("%v\n", msSecond)
		}
		usTime := `answer{} 42 ` + fmt.Sprintf("%v\n", usStamp)
		nsTime := `answer{} 42 ` + fmt.Sprintf("%v\n", nsStamp)
		sTime := `answer{} 42 ` + fmt.Sprintf("%v\n", sStamp)
		assert.Equal(t, msTime, dataString)
		assert.NotEqual(t, usTime, dataString)
		assert.NotEqual(t, nsTime, dataString)
		assert.NotEqual(t, sTime, dataString)
	}
}

func TestConvertMetricWithNilMetrics(t *testing.T) {
	event := corev2.FixtureEvent("entity1", "check1")
	event.Check = nil
	event.Metrics = nil
	_, err := convertMetrics(event)
	assert.NoError(t, err)
}
func TestConvertMetricWithNilMetricsPoints(t *testing.T) {
	event := corev2.FixtureEvent("entity1", "check1")
	event.Check = nil
	event.Metrics = corev2.FixtureMetrics()
	event.Metrics.Points = nil
	_, err := convertMetrics(event)
	assert.NoError(t, err)
}

func TestSendMetrics(t *testing.T) {
	event := corev2.FixtureEvent("entity1", "check1")
	event.Check = nil
	event.Metrics = corev2.FixtureMetrics()
	msStamp := int64(1624376039373)
	nsStamp := int64(1624376039373111122)
	msTime := `answer{foo="bar"} 42 ` + fmt.Sprintf("%v", msStamp)
	for _, p := range event.Metrics.Points {
		p.Timestamp = nsStamp
	}

	var test = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		body, err := ioutil.ReadAll(r.Body)
		assert.NoError(t, err)
		expectedBody := msTime
		assert.Equal(t, expectedBody, strings.Trim(string(body), "\n"))
		w.WriteHeader(http.StatusOK)
	}))

	url, err := url.ParseRequestURI(test.URL)
	assert.NoError(t, err)
	plugin.Url = url.String()
	dataString, err := convertMetrics(event)
	assert.NoError(t, err)
	assert.NoError(t, sendMetrics(dataString))
}
func TestSendMetricsDryRun(t *testing.T) {
	plugin.DryRun = true
	event := corev2.FixtureEvent("entity1", "check1")
	event.Check = nil
	event.Metrics = corev2.FixtureMetrics()
	msStamp := int64(1624376039373)
	nsStamp := int64(1624376039373111122)
	msTime := `answer{foo="bar"} 42 ` + fmt.Sprintf("%v", msStamp)
	for _, p := range event.Metrics.Points {
		p.Timestamp = nsStamp
	}

	var test = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		body, err := ioutil.ReadAll(r.Body)
		assert.NoError(t, err)
		expectedBody := msTime
		assert.Equal(t, expectedBody, strings.Trim(string(body), "\n"))
		w.WriteHeader(http.StatusOK)
	}))

	url, err := url.ParseRequestURI(test.URL)
	assert.NoError(t, err)
	plugin.Url = url.String()
	dataString, err := convertMetrics(event)
	assert.NoError(t, err)
	assert.NoError(t, sendMetrics(dataString))
	clearPlugin()
}

func TestSendLog(t *testing.T) {
	event := corev2.FixtureEvent("entity1", "check1")
	event.Check = nil
	event.Metrics = nil
	msgBytes, err := json.Marshal(event)
	assert.NoError(t, err)

	var test = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		body, err := ioutil.ReadAll(r.Body)
		assert.NoError(t, err)
		expectedBody := string(msgBytes)
		assert.Equal(t, expectedBody, strings.Trim(string(body), "\n"))
		w.WriteHeader(http.StatusOK)
	}))

	url, err := url.ParseRequestURI(test.URL)
	assert.NoError(t, err)
	plugin.Url = url.String()
	assert.NoError(t, sendLog(string(msgBytes)))
}
func TestSendLogDryRun(t *testing.T) {
	plugin.DryRun = true
	event := corev2.FixtureEvent("entity1", "check1")
	event.Check = nil
	event.Metrics = nil
	msgBytes, err := json.Marshal(event)
	assert.NoError(t, err)

	var test = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		body, err := ioutil.ReadAll(r.Body)
		assert.NoError(t, err)
		expectedBody := string(msgBytes)
		assert.Equal(t, expectedBody, strings.Trim(string(body), "\n"))
		w.WriteHeader(http.StatusOK)
	}))

	url, err := url.ParseRequestURI(test.URL)
	assert.NoError(t, err)
	plugin.Url = url.String()
	assert.NoError(t, sendLog(string(msgBytes)))
	clearPlugin()
}

func TestMsTimestamp(t *testing.T) {
	event := corev2.FixtureEvent("entity1", "check1")
	event.Metrics = corev2.FixtureMetrics()
	msStamp := int64(1624376039373)
	nsStamp := int64(1624376039373111122)
	secStamp := int64(1624376039)
	expectedStamp := msStamp
	finalStamp := msTimestamp(nsStamp)
	assert.Equal(t, expectedStamp, finalStamp)
	expectedStamp = secStamp * 1000
	finalStamp = msTimestamp(secStamp)
	assert.Equal(t, expectedStamp, finalStamp)
}

func TestExecuteHandler(t *testing.T) {
	plugin.MetricDimensions = `hey=now,this=that`
	plugin.MetricMetadata = `you=me,here=there`
	plugin.LogFields = `near=far,in=out`
	plugin.SourceNameTemplate = defaultNameTemplate
	plugin.SourceHostTemplate = defaultHostTemplate
	plugin.SourceCategoryTemplate = defaultCategoryTemplate

	event := corev2.FixtureEvent("entity1", "check1")
	event.Metrics = corev2.FixtureMetrics()
	msStamp := int64(1624376039373)
	nsStamp := int64(1624376039373111122)
	msTime := `answer{foo="bar"} 42 ` + fmt.Sprintf("%v", msStamp)
	for _, p := range event.Metrics.Points {
		p.Timestamp = nsStamp
	}
	event.Timestamp = msTimestamp(event.Timestamp)
	logMsg, err := createLogMsg(event)
	assert.NoError(t, err)
	expectedBytes, err := json.Marshal(logMsg)
	assert.NoError(t, err)
	plugin.EnableSendLog = true
	plugin.EnableSendMetrics = true
	results := 0
	var test = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		body, err := ioutil.ReadAll(r.Body)
		assert.NoError(t, err)
		switch {
		case contains(r.Header["Content-Type"], "application/vnd.sumologic.prometheus"):
			// recieved metrics with Content-Type header set
			expectedBody := msTime
			assert.Equal(t, expectedBody, strings.Trim(string(body), "\n"))
			assert.Equal(t, plugin.MetricDimensions, r.Header["X-Sumo-Dimensions"][0])
			assert.Equal(t, plugin.MetricMetadata, r.Header["X-Sumo-Metadata"][0])
			results++
		case contains(r.Header["Content-Type"], "application/json"):
			// recieved log with Content-Type header unset
			expectedBody := string(expectedBytes)
			assert.Equal(t, expectedBody, strings.Trim(string(body), "\n"))
			assert.Equal(t, plugin.LogFields, r.Header["X-Sumo-Fields"][0])
			results++
		default:
			assert.FailNow(t, "No Content-Type Header")

		}
		if len(plugin.SourceNameTemplate) > 0 {
			assert.Equal(t, plugin.SourceName, r.Header["X-Sumo-Name"][0])
		} else {
			assert.Nil(t, r.Header["X-Sumo-Name"])
		}
		if len(plugin.SourceHostTemplate) > 0 {
			assert.Equal(t, plugin.SourceHost, r.Header["X-Sumo-Host"][0])
		} else {
			assert.Nil(t, r.Header["X-Sumo-Host"])
		}
		if len(plugin.SourceCategoryTemplate) > 0 {
			assert.Equal(t, plugin.SourceCategory, r.Header["X-Sumo-Category"][0])
		} else {
			assert.Nil(t, r.Header["X-Sumo-Category"])
		}
		w.WriteHeader(http.StatusOK)
	}))

	url, err := url.ParseRequestURI(test.URL)
	assert.NoError(t, err)
	plugin.Url = url.String()
	err = executeHandler(event)
	assert.Equal(t, results, 2)
	assert.NoError(t, err)
}


main.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"math"
	"net/http"
	"time"

	corev2 "github.com/sensu/sensu-go/api/core/v2"
	"github.com/sensu/sensu-plugin-sdk/sensu"
	"github.com/sensu/sensu-plugin-sdk/templates"
)

// Config represents the handler plugin config.
type Config struct {
	sensu.PluginConfig
	Url                    string
	Verbose                bool
	DryRun                 bool
	EnableSendLog          bool
	EnableSendMetrics      bool
	SourceName             string
	SourceNameTemplate     string
	SourceHost             string
	SourceHostTemplate     string
	SourceCategory         string
	SourceCategoryTemplate string
	MetricDimensions       string
	MetricMetadata         string
	LogFields              string
}
type LogMsg struct {
	Data []interface{} `json:"data"`
}

const (
	defaultHostTemplate     = "{{ .Entity.Name }}"
	defaultNameTemplate     = "{{ .Check.Name }}"
	defaultCategoryTemplate = "sensu-event"
)

var (
	plugin = Config{
		PluginConfig: sensu.PluginConfig{
			Name:     "sensu-sumologic-handler",
			Short:    "Send Sensu observability data (events and metrics) to a hosted Sumo Logic HTTP Logs and Metrics Source.",
			Keyspace: "sensu.io/plugins/sumologic/config",
		},
	}
	options = []*sensu.PluginConfigOption{
		&sensu.PluginConfigOption{
			Path:      "url",
			Env:       "SUMOLOGIC_URL",
			Argument:  "url",
			Shorthand: "u",
			Default:   "",
			Usage:     "Sumo Logic HTTP Logs and Metrics Source URL (Required)",
			Secret:    true,
			Value:     &plugin.Url,
		},
		&sensu.PluginConfigOption{
			Path:      "verbose",
			Argument:  "verbose",
			Shorthand: "v",
			Default:   false,
			Usage:     "Verbose output to stdout",
			Value:     &plugin.Verbose,
		},
		&sensu.PluginConfigOption{
			Path:      "dry-run",
			Argument:  "dry-run",
			Shorthand: "n",
			Default:   false,
			Usage:     "Dry-run, do not send data to Sumo Logic collector, report to stdout instead",
			Value:     &plugin.DryRun,
		},
		&sensu.PluginConfigOption{
			Path:      "send-log",
			Env:       "SUMOLOGIC_SEND_LOG",
			Argument:  "send-log",
			Shorthand: "l",
			Default:   false,
			Usage:     "Send event as log",
			Value:     &plugin.EnableSendLog,
		},
		&sensu.PluginConfigOption{
			Path:      "send-metrics",
			Env:       "SUMOLOGIC_SEND_METRICS",
			Argument:  "send-metrics",
			Shorthand: "m",
			Default:   false,
			Usage:     "Send event metrics, if there are metrics attached to sensu event",
			Value:     &plugin.EnableSendMetrics,
		},
		&sensu.PluginConfigOption{
			Path:     "source-name",
			Env:      "SUMOLOGIC_SOURCE_NAME",
			Argument: "source-name",
			Default:  defaultNameTemplate,
			Usage:    "Custom Sumo Logic source name (supports handler templates)",
			Value:    &plugin.SourceNameTemplate,
		},
		&sensu.PluginConfigOption{
			Path:     "source-host",
			Env:      "SUMOLOGIC_SOURCE_HOST",
			Argument: "source-host",
			Default:  defaultHostTemplate,
			Usage:    "Custom Sumo Logic source host (supports handler templates)",
			Value:    &plugin.SourceHostTemplate,
		},
		&sensu.PluginConfigOption{
			Path:     "source-category",
			Env:      "SUMOLOGIC_SOURCE_CATEGORY",
			Argument: "source-category",
			Default:  defaultCategoryTemplate,
			Usage:    "Custom Sumo Logic source category (supports handler templates)",
			Value:    &plugin.SourceCategoryTemplate,
		},
		&sensu.PluginConfigOption{
			Path:     "metric-dimensions",
			Env:      "SUMOLOGIC_METRIC_DIMENSIONS",
			Argument: "metric-dimensions",
			Default:  "",
			Usage:    "Custom Sumo Logic metric dimensions (comma separated key=value pairs)",
			Value:    &plugin.MetricDimensions,
		},
		/* JDS: metric metadata is being deprecated in the sumo http source in favor of metric dimensions
		&sensu.PluginConfigOption{
			Path:     "metric-metadata",
			Env:      "SUMOLOGIC_METRIC_METADATA",
			Argument: "metric-metadata",
			Default:  "",
			Usage:    "Custom Sumo Logic metric metadata (comma separated key=value pairs)",
			Value:    &plugin.MetricMetadata,
		},
		*/
		&sensu.PluginConfigOption{
			Path:     "log-fields",
			Env:      "SUMOLOGIC_LOG_FIELDS",
			Argument: "log-fields",
			Default:  "",
			Usage:    "Custom Sumo Logic log fields (comma separated key=value pairs)",
			Value:    &plugin.LogFields,
		},
	}
)

func main() {
	handler := sensu.NewGoHandler(&plugin.PluginConfig, options, checkArgs, executeHandler)
	handler.Execute()
}

func checkArgs(event *corev2.Event) error {
	if !plugin.EnableSendMetrics && !plugin.EnableSendLog {
		return fmt.Errorf("Must have at least one of --send-log or --send-metrics")
	}
	if len(plugin.Url) == 0 {
		return fmt.Errorf("--url or SUMOLOGIC_URL environment variable is required")
	}
	if plugin.DryRun {
		plugin.Verbose = true
	}
	return nil
}

func executeHandler(event *corev2.Event) error {
	err := renderTemplates(event)
	if err != nil {
		log.Printf("Error rendering templates: %s", err)
	}

	dataString, err := convertMetrics(event)
	if err != nil {
		return err
	}
	doMetrics := false
	if plugin.EnableSendMetrics && len(dataString) > 0 {
		doMetrics = true
	}
	if plugin.Verbose && plugin.EnableSendMetrics && len(dataString) == 0 {
		log.Printf("Warning: metrics sending enabled, but no metrics found in Sensu event")
	}

	doLog := plugin.EnableSendLog

	if plugin.Verbose {
		log.Printf("Info: Sending Metrics: %v Sending Log: %v",
			doMetrics, doLog)
	}

	if doMetrics {
		err = sendMetrics(dataString)
		if err != nil {
			return err
		}
	}

	if doLog {
		logMsg, err := createLogMsg(event)
		if err != nil {
			return err
		}
		msgBytes, err := json.Marshal(logMsg)
		if err != nil {
			return err
		}

		err = sendLog(string(msgBytes))
		if err != nil {
			return err
		}
	}

	return nil
}

func createLogMsg(event *corev2.Event) (LogMsg, error) {
	timestamp := msTimestamp(event.Timestamp)
	logMsg := LogMsg{}
	t := make(map[string]int64)
	e := make(map[string]*corev2.Event)
	t["timestamp"] = timestamp
	e["event"] = event
	logMsg.Data = append(logMsg.Data, t)
	logMsg.Data = append(logMsg.Data, e)
	return logMsg, nil

}

func renderTemplates(event *corev2.Event) error {
	if len(plugin.SourceHostTemplate) > 0 {
		sourceHost, err := templates.EvalTemplate("source-host", plugin.SourceHostTemplate, event)
		if err != nil {
			return fmt.Errorf("%s: Error processing source host template: %s Err: %s",
				plugin.PluginConfig.Name, plugin.SourceHostTemplate, err)
		}
		plugin.SourceHost = sourceHost
	}
	if len(plugin.SourceNameTemplate) > 0 {
		sourceName, err := templates.EvalTemplate("source-name", plugin.SourceNameTemplate, event)
		if err != nil {
			return fmt.Errorf("%s: Error processing source name template: %s Err: %s",
				plugin.PluginConfig.Name, plugin.SourceNameTemplate, err)
		}
		plugin.SourceName = sourceName
	}
	if len(plugin.SourceCategoryTemplate) > 0 {

		sourceCategory, err := templates.EvalTemplate("source-category", plugin.SourceCategoryTemplate, event)
		if err != nil {
			return fmt.Errorf("%s: Error processing source category template: %s Err: %s",
				plugin.PluginConfig.Name, plugin.SourceCategoryTemplate, err)
		}
		plugin.SourceCategory = sourceCategory
	}
	return nil
}

func msTimestamp(ts int64) int64 {
	/* Auto detection of metric point timestamp precision using a heuristic with a 250-ish year cutoff */
	timestamp := ts
	switch ts := math.Log10(float64(timestamp)); {
	case ts < 10:
		// assume timestamp is seconds convert to millisecond
		timestamp = time.Unix(timestamp, 0).UnixNano() / int64(time.Millisecond)
	case ts < 13:
		// assume timestamp is milliseconds
	case ts < 16:
		// assume timestamp is microseconds
		timestamp = (timestamp * 1000) / int64(time.Millisecond)
	default:
		// assume timestamp is nanoseconds
		timestamp = timestamp / int64(time.Millisecond)

	}
	return timestamp

}

func convertMetrics(event *corev2.Event) (string, error) {
	output := ""
	if event.Metrics != nil {
		for _, point := range event.Metrics.Points {
			tags := ""
			for i, tag := range point.Tags {
				if len(point.Tags)-1 == i {
					tags = tags + fmt.Sprintf("%s=\"%v\"", tag.Name, tag.Value)
				} else {
					tags = tags + fmt.Sprintf("%s=\"%v\", ", tag.Name, tag.Value)
				}
			}
			timestamp := msTimestamp(point.Timestamp)
			output += fmt.Sprintf("%s{%s} %v %v\n", point.Name, tags, point.Value, timestamp)
		}
	}
	return output, nil
}

func sendMetrics(dataString string) error {
	client := &http.Client{}
	req, err := http.NewRequest("POST", plugin.Url, bytes.NewBufferString(dataString))
	if err != nil {
		return fmt.Errorf("New Http Request failed: %s", err)
	}
	req.Header.Add(`Content-Type`, "application/vnd.sumologic.prometheus")
	// Add optional headers here
	if len(plugin.SourceHost) > 0 {
		req.Header.Add(`X-Sumo-Host`, plugin.SourceHost)
	}
	if len(plugin.SourceName) > 0 {
		req.Header.Add(`X-Sumo-Name`, plugin.SourceName)
	}
	if len(plugin.SourceCategory) > 0 {
		req.Header.Add(`X-Sumo-Category`, plugin.SourceCategory)
	}
	if len(plugin.MetricDimensions) > 0 {
		req.Header.Add(`X-Sumo-Dimensions`, plugin.MetricDimensions)
	}
	if len(plugin.MetricMetadata) > 0 {
		req.Header.Add(`X-Sumo-Metadata`, plugin.MetricMetadata)
	}

	// If DryRun report back request details
	if plugin.DryRun {
		bytes, _ := ioutil.ReadAll(req.Body)
		fmt.Printf("Dry Run Metric Request:  \n Method: %v Url: %v\n Headers: %+v\n Data:\n%v\n",
			req.Method, req.URL, req.Header, string(bytes))
		return nil
	}

	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("POST metrics to %s failed: %s", plugin.Url, err)
	}

	if resp.StatusCode < 200 || resp.StatusCode > 299 {
		return fmt.Errorf("POST metrics to %s failed with status %v", plugin.Url, resp.Status)
	}

	defer resp.Body.Close()

	return nil
}
func sendLog(dataString string) error {
	client := &http.Client{}
	req, err := http.NewRequest("POST", plugin.Url, bytes.NewBufferString(dataString))
	if err != nil {
		return fmt.Errorf("New Http Request failed: %s", err)
	}
	req.Header.Add(`Content-Type`, "application/json")
	// Add optional headers here
	if len(plugin.SourceHost) > 0 {
		req.Header.Add(`X-Sumo-Host`, plugin.SourceHost)
	}
	if len(plugin.SourceName) > 0 {
		req.Header.Add(`X-Sumo-Name`, plugin.SourceName)
	}
	if len(plugin.SourceCategory) > 0 {
		req.Header.Add(`X-Sumo-Category`, plugin.SourceCategory)
	}
	if len(plugin.LogFields) > 0 {
		req.Header.Add(`X-Sumo-Fields`, plugin.LogFields)
	}

	// If DryRun report back request details
	if plugin.DryRun {
		bytes, _ := ioutil.ReadAll(req.Body)
		fmt.Printf("Dry Run Log Request:  \n Method: %v Url: %v\n Headers: %+v\n Data:\n%v\n",
			req.Method, req.URL, req.Header, string(bytes))
		return nil
	}

	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("POST log to %s failed: %s", plugin.Url, err)
	}

	if resp.StatusCode < 200 || resp.StatusCode > 299 {
		return fmt.Errorf("POST log to %s failed with status %v", plugin.Url, resp.Status)
	}

	defer resp.Body.Close()

	return nil
}
